keps/sig-api-machinery/1623-standardize-conditions/README.md

```diff
@@ -0,0 +1,144 @@
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+While many Kuberentes APIs have `.status.conditions`, the schema of `condition` varies a lot between them.
+There is very little commonality at the level of serialization, proto-encoding, and required vs optional.
+Conditions are central enough to the API to make a common golang type with a fixed schema.
+The schema can be a strong recommendation to all API authors.
+
+## Motivation
+
+Allow general consumers to expect a common schema for `.status.conditions` and share golang logic for common Get, Set, Is for `.status.conditions`.
+The pattern is well-established and we have a good sense of the schema we now want.
+
+### Goals
+
+ 1. For all new APIs, have a common type for `.status.conditions`.
+ 2. Provide common utility methods for `HasCondition`, `IsConditionTrue`, `SetCondition`, etc.
+ 3. Provide recommended defaulting functions that set required fields and can be embedded into conversion/default functions.
+
+### Non-Goals
+
+ 1. Update all existing APIs to make use of the new condition type.
+
+## Proposal
+
+Introduce a type into k8s.io/apimachinery/pkg/apis/meta/v1 for `Condition` that looks like 
+```go
+type Condition struct {
+	// Type of condition in CamelCase.
```

Comment:
We've documented CamelCase, but we've also told people to use conditions as an extension mechanism.  To do that safely, we should probably support or endorse company.com/label style names...

Reply:
Do we expect multiple controllers to be updating the same object's `status` with different Conditions?

If we assume that a single controller owns the `status` of the object, then I don't think we need namespaced labels for the types (since the controller should be able to ensure uniqueness).

If we expect multiple controllers to be updating `status`, then this is more of a problem -- I'd be concerned about other implications (controller fighting) in that design, though.

Reply:
> Do we expect multiple controllers to be updating the same object's status with different Conditions?

Yes. NPD, kubelet, and the node controller all set node conditions.

> If we expect multiple controllers to be updating status, then this is more of a problem -- I'd be concerned about other implications (controller fighting) in that design, though.

The entire k8s API system is about making it safe for humans and controllers (and multiple controllers) work together on particular objects, both on their status and spec.

Reply:
A while back we documented Conditions a bit more.  The use of conditions has evolved a lot from "a single controller" to "how 3rd parties add extension status".

Case in point - pod `readinessGates` expect conditions to be populated to indicate "readiness" from an external POV (eg. LB programming).

I raised the prefixing at that time, but we didn't really resolve it.

Reply:
> We've documented CamelCase, but we've also told people to use conditions as an extension mechanism. To do that safely, we should probably support or endorse company.com/label style names...

That seems reasonable.  `company.com/CamelCase` and `CamelCase` for first-class conditions?  I do think that some are realistically more first class than others.

Reply:
Define "first-class" ?

labels are GENERALLY not using CamelCase, but in this case I won't balk.